### DISCUSSION TITLE
Implementing a query directive to remove fields from the response

### DISCUSSION THREAD

**Discussion by Aqueelmiq** [graphql-java/graphql-java#3044] [Q&A]
Hi, we are implementing a new directive at Twitter in GraphQL java. We are trying to build a variant of `@skip` directive which takes in a value instead of a simple yes/no boolean argument. The goal here is to exclude the field from the response if the value of the field returned by the data fetcher for the field is the same as the value provided in the directive argument.

Here is the SDL

```
"allows for fields to be excluded if it matches a provided value"
directive @exclude(
    """
    Excludes the field from the query if the field response matches the value provided to this argument.
    """
    if: String!
  ) on FIELD
```

Example usage

```
# request
home {
  tweet {
    user {
      is_verified 
      coordinates
    }
  }
}

# response 1
"home": {
  [
    "tweet": {
      "user": {
        "is_verified": true,
         "coordinates": "0000"
      }
    }
  ]
  
# request
home {
  tweet {
    user {
      is_verified 
      coordinates @exclude(if: “0000”)
    }
  }
}

# response 2
"home": {
  [
    "tweet": {
      "user": {
        "is_verified": true
      }
    }
  ]
```

Now let's say the coordinates field on user is required

```
type User {
  is_verified: Boolean
  coordinates: String!
}
```

if I implement this directive and return a null from the data fetcher for coordinates field when the value matches the directive value. I get an exception "NonNullableFieldWasNullException". This makes sense, I have violated GraphQL spec here. However, this was explicitly requested by the client, so I want to introduce this behavior in a way that we remove the field from selection set post resolution of the operation.

How can I implement this reasonably in GraphQL java?

**Comment by bbakerman** [comment]
This feels like a post processing thing.  That is you want to break the rules of graphql (around non nullability) - so trying to get the engine to do it is tricky - since it wants to enforce the rules.

Have a look at `graphql.execution.instrumentation.Instrumentation#instrumentExecutionResult(graphql.ExecutionResult, graphql.execution.instrumentation.parameters.InstrumentationExecutionParameters, graphql.execution.instrumentation.InstrumentationState)`

It allows you to change the result object and it runs pretty much at the end of the engine run.

In theory you can change the result object to anything at all

```java
            //
            // allow instrumentation to tweak the result
            executionResult = executionResult.thenCompose(result -> instrumentation.instrumentExecutionResult(result, instrumentationParameters, instrumentationState));
```

The challenge you will have is that a graphql is a complicated mechanism - walking that result in terms of the "query" and types and following the rules system is EXACTLY what the graphql-engine is for

However assumming that there are no parsing and validation errors in the execution result its possible to do

You can use other `Instrumentation` methods to stash away "execution related" objects - eg the parsed query document or even the `graphql.execution.ExecutionContext`

**Reply by Aqueelmiq** [reply]
> That is you want to break the rules of graphql (around non nullability)

I think one way to look at this behaviour we want to implement is we want to remove the field from the query execution after the query is executed and with the knowledge of the value resolved for every field. So it is essentially @skip directive implemented post execution if that makes sense.

> The challenge you will have is that a graphql is a complicated mechanism - walking that result in terms of the "query" and types and following the rules system is EXACTLY what the graphql-engine is for

Yeah that's complicated and will probably come with a performance penalty if we traverse the full object again. is there no way to remove the field from query selection set after we resolved a value for that field?

**Comment by bbakerman** [comment]
>  is there no way to remove the field from query selection set after we resolved a value for that field?

In short no - we dont have a callback (after a `DataFetcher`) that allows you to change the field result or remove it

in a graphql design sense what you are asking regarding non nulls is strange.

That is `coordinates: String!` says to a caller "the `coordinates` field is a `String` and its ALWAYS non null" and yet you want it to be "the `coordinates` field is a `String` and while it says its ALWAYS non null - sometimes its not"
